File: 0071-simplify-path/0071-simplify-path.cpp
class Solution {
public:
    string simplifyPath(string path) {
        stack<string> st;
        string temp = "";
        for(char c : path){
            if(c == '/' && temp != ""){
                if(temp == ".."){
                    if(!st.empty())
                        st.pop();
                }
                else if(temp != "."){
                    st.push(temp);
                }
                temp = "";
            }
            else if(c != '/') temp += c;
        }
        if(temp == ".."){
            if(!st.empty())
                st.pop();
        }
        else if(temp != "" && temp != "."){
            st.push(temp);
        }
        string result = "";
        while(!st.empty()){
            string s = st.top();
            st.pop();
            result = ("/" + s) + result;
        }
        if(result == "") result = "/";
        return result;
    }
};

File: 0071-simplify-path/0071-simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        Stack<String> st = new Stack<>();
        for (String part : path.split("/")) {
            if (part.isEmpty() || part.equals(".")) {
                continue;
            }
            if (part.equals("..")) {
                if (!st.isEmpty()) {
                    st.pop();
                }
            } else {
                st.push(part);
            }
        }
        StringBuilder res = new StringBuilder();

        while (!st.isEmpty()) {
            res.insert(0, "/" + st.pop());
        }
        return res.length() == 0 ? "/" : res.toString();
    }
}